Allow listed add-ons to become unlisted (bug 1122108)

robhudson opened the pull request.

robhudson commented on apps/devhub/templates/devhub/versions/list.html:
```diff
@@ -277,6 +277,42 @@ <h3 data-tmpl="{{ _('Delete Version {version}') }}"></h3>
+          public facing page, and won't be updated automatically for current
+          users.  It is recommended for unlisted add-ons to provide a custom
+          <updateURL> (link to mdn) in their install.rdf file for automatic
+          updates.
```
I need to update this to include the real URL.

robhudson commented:
@magopian This is how far I got by end of day Friday. If you wouldn't mind giving it a pre-review and I'll get back on this on Monday to fix the text and also add any edge case stuff that needs to go in. Thanks.

magopian commented on apps/devhub/templates/devhub/versions/list.html:
```diff
@@ -277,6 +277,42 @@ <h3 data-tmpl="{{ _('Delete Version {version}') }}"></h3>
+        {% endtrans %}
+      </p>
+      {{ csrf() }}
+      <input type="hidden" name="version_id" class="version_id">
```
I don't think we need the version (same for the `$('.version_id', this).val($(d.click_target).attr('data-version'));` in the modal declared below in the devhub.js file)

magopian commented on apps/devhub/tests/test_views_versions.py:
```diff
@@ -186,6 +187,20 @@ def test_user_can_disable_addon(self, hide_mock):
+        self.addon.update(status=amo.STATUS_PUBLIC, disabled_by_user=False,
+                          is_listed=True)
+        res = self.client.post(self.unlist_url)
+        eq_(res.status_code, 302)
```
I try to write all the new test code using the "pytest" way (simple assertions). What do you think?

robhudson replied:
I could do that. I was mostly following convention here.

magopian replied:
That the difficult part: striking a balance between convention/coherence and the "better way". I try to use the pytest way whenever I can, meaning whenever I'm writing a new test (or rewriting most of an old one, in which case i rewrite it entirely).

It's open to discussion though.

magopian commented on apps/devhub/tests/test_views_versions.py:
```diff
@@ -267,7 +283,9 @@ def test_hidden_addon_radio(self):
         assert not doc('.unlist-addon').attr('checked')
-        assert doc('.unlist-addon').attr('disabled') == 'disabled'  # Readonly.
+        assert not doc('.unlist-addon').attr('disabled')
+        assert not doc('#modal-disable')
```
:+1:

magopian commented on static/js/zamboni/devhub.js:
```diff
@@ -226,6 +226,18 @@ $(document).ready(function() {
+    if ($("#modal-unlist").length) {
+        $modalDisable = $('#modal-unlist').modal('.unlist-addon', {
+            width: 400,
+            callback: function(d){
```
I don't think we need a callback at all here

magopian commented:
This is looking nice!

I think it also misses the "(soft) delete the user reviews" part detailed in the bug.

Did you test if the addon is indeed unlisted after the `is_listed` field is updated to `False`? Don't we need to reindex, or clean some cache or something? We must make sure that the addon doesn't have a public page anymore, nor a "versions" page, that it's not listed anywhere (search results, developer's page, reviews...).

robhudson commented:
Updated:

* I fixed the commented issues
* I added an unlisted exclusion to the `reindex` command
* Upon unlisting the add-on is now removed from the search index

In comment 4 of the bug it mentions, "If there are versions awaiting review, they can either be disabled, or put through the normal automated validation/manual review process. Either is fine with me."  Do you have a preference? Or advice on how this might be done?

magopian commented:
Regarding your last point: changing the `is_listed` field on the addon should already do the right thing regarding listing the versions waiting to be reviewed in the appropriate unlisted review queues.

I don't think we need to disable the versions.

magopian commented on apps/devhub/templates/devhub/versions/list.html:
```diff
@@ -277,6 +277,40 @@ <h3 data-tmpl="{{ _('Delete Version {version}') }}"></h3>
+        {{ _('Unlist Add-on') }}
+      </h3>
+      <p>
+        {% trans update_url='https://developer.mozilla.org/Add-ons/Install_Manifests#updateURL' %}
```
Using this url automatically redirects me to the french localized version of the page by prepending a `fr`. However, this page doesn't exist in french, so I get a 404.

Maybe a better idea would be to use the `en-US` locale, and then users will be able to change that once on the page using MDN's locale selector?

magopian commented on apps/devhub/tests/test_views_versions.py:
```diff
@@ -186,6 +187,24 @@ def test_user_can_disable_addon(self, hide_mock):
+        self.addon.update(status=amo.STATUS_PUBLIC, disabled_by_user=False,
+                          is_listed=True)
+        res = self.client.post(self.unlist_url)
+        self.assertEqual(res.status_code, 302)
```
I'm sorry, I think I wasn't clear enough when I was talking about using the "pytest way": I meant using plain `assert` instead of helpers like `eq_` or `self.assert*`. It's what's done below in `test_non_owner_cant_change_status` for example:

```python
assert res.status_code == 302
...
assert addon.status == amo.STATUS_PUBLIC
assert not addon.is_listed
...
```

robhudson replied:
Yeah, I wasn't clear on that. But, I'm curious why? The `self.assert*` methods are part of Python's `unittest` lib, not nose. And django builds upon that and adds a few of their own that are useful.

magopian replied:
I'm not against those, it's just that they're often useless like the `self.assertEqual` that doesn't do more, and often less, than py.test's `assert x == y`, and I find them un-pythonic (camelcase? really?), often more verbose, and tie the tests to their class.

The last point is moot because we're not going to rewrite all our tests to be simple functions using pytest fixtures.

Some are handy, like `assertTemplateUsed`, or `assertFormError`, but they're rarely used.

To sum it up: I think pytest's assertions introspections are more powerful (see the message you have when comparing two dicts or tuples, telling you which elements are missing or added, or which keys have different values...), and they look more pythonic, that's why I try to use them whenever I can.

What do you think?

robhudson replied:
All excellent points. I didn't realize py.test overrode the builtin `assert`. That does sound useful.

magopian replied:
You've modified the rest of the test, but it seems you missed this line:
`assert res.status_code == 302`

magopian commented on apps/addons/cron.py:
```diff
@@ -454,6 +454,7 @@ def reindex_addons(index=None, addon_type=None):
     ids = (Addon.objects.values_list('id', flat=True)
            .filter(_current_version__isnull=False,
                    status__in=amo.VALID_STATUSES,
+                   is_listed=True,
```
:+1:

magopian replied:
Do we also need this in other places like in `addons.tasks.index_addons`?

When we submit a new addon, where/how is it index? I think we need to do something similar there so unlisted addons don't get indexed at all (same for versions and files?).

This could obviously be done in another PR, do you want me to open a new bug?

robhudson replied:
We may need it in more places. I'll take a look. Probably the `index_addons` could short circuit if the addon is not listed.

magopian commented:
This is excellent, thanks!
r+wc

robhudson commented:
Updated:

* I fixed the test assertions
* I added the correct URL for MDN so the locale redirect works
* I added a filter to the `index_addons`. This may have some test failures if we relied on tests not caring about status or disabled but Travis was having problems.

magopian commented on apps/devhub/tests/test_views_versions.py:
```diff
@@ -186,6 +187,24 @@ def test_user_can_disable_addon(self, hide_mock):
+        self.assertEqual(res.status_code, 302)
+        addon = Addon.with_unlisted.get(id=3615)
+        assert addon.status == amo.STATUS_PUBLIC
+        assert addon.is_listed == False
```
That should either
`assert not addon.is_listed`
or
`assert addon.is_listed is False`

I usually go with the first one.

(Flake8 complains about that in travis btw).

magopian commented on apps/devhub/tests/test_views_versions.py:
```diff
@@ -186,6 +187,24 @@ def test_user_can_disable_addon(self, hide_mock):
+        entry = ActivityLog.objects.get()
+        assert entry.action == amo.LOG.ADDON_UNLISTED.id
+        msg = entry.to_string()
+        assert self.addon.name.__unicode__() in msg, ("Unexpected: %r" % msg)
```
You can of course provide the message part of the assert, but I usually don't bother, because pytest does the right thing and shows you what it was expecting, and what it got instead ;)

magopian commented on apps/addons/models.py:
```diff
@@ -1578,7 +1578,9 @@ def version_changed(sender, **kw):
-        tasks.index_addons.delay([instance.id])
+        if (not instance.disabled_by_user and instance.is_listed and
+                instance.status in amo.VALID_STATUSES):
+            tasks.index_addons.delay([instance.id])
```
Nice one ;) It's strange we didn't have this before, does that mean we used to have user disabled or incomplete addons show up in the searches?

magopian commented:
There's a few travis failures caused by the changes in the way addons are indexed: https://travis-ci.org/mozilla/olympia/jobs/59458324#L458

I think most of those could be simply fixed by setting the correct status on the addons.

It's still time to split of those changes (related to indexing) in another PR if you feel it's better.

robhudson commented:
Updated based on last review. Also squashed and got rid of the search indexing changes.

magopian commented on apps/devhub/templates/devhub/versions/list.html:
```diff
@@ -277,6 +277,40 @@ <h3 data-tmpl="{{ _('Delete Version {version}') }}"></h3>
+        {{ _('Unlist Add-on') }}
+      </h3>
+      <p>
+        {% trans update_url='https://developer.mozilla.org/docs/Mozilla/Add-ons/Install_Manifests#updateURL' %}
```
This works well now :+1:

magopian commented:
Looks perfect, thanks!
Excellent idea to split out the search indexing changes, it'll be easier to follow ;)

r+
